python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import weakref
+from collections import defaultdict
+from typing import Any, DefaultDict, Tuple, Type, TypeVar
+from weakref import WeakSet
+
+import rmm
+
+from cudf.core.buffer.buffer import Buffer, cuda_array_interface_wrapper
+
+T = TypeVar("T", bound="CopyOnWriteBuffer")
+
+
+def _keys_cleanup(ptr, size):
+    weak_set_values = CopyOnWriteBuffer._instances[(ptr, size)]
+    if (
+        len(weak_set_values) == 1
+        and next(iter(weak_set_values.data))() is None
+    ):
+        # When the last remaining reference is being cleaned up we will still
+        # have a dead weak-reference in `weak_set_values`, if that is the case
+        # we are good to perform the key's cleanup
+        del CopyOnWriteBuffer._instances[(ptr, size)]
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        key = (self._ptr, self._size)
+        self.__class__._instances[key].add(self)
+        self._zero_copied = False
+        weakref.finalize(self, _keys_cleanup, self._ptr, self._size)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
```

Comment:
The above case works and the reverse too works. We achieve this by making a copy while CAI's are accessed on a weakly reference buffer.

Reply:
Hmm when I originally wrote this comment I had assumed that the reverse wouldn't work because accessing the CAI automatically made a copy. However, I think now I'm seeing where I misunderstood. Are my example and the notes in it correct?
```
>>> x = cudf.Series([1, 2, 3])
>>> y = x[:]  # This will be a view into the same data 
>>> a = cupy.asarray(y) # This will trigger `y` to unlink from `x`, so `y` is now a deep copy of `x`, but `a` is a view of `y`. Is that right?
```

Reply:
> but `a` is a view of `y`

That is how I understand it. You still get a view of the thing you exposed to the outside world, it's just it is unlinked from any internal views (by copying) before you do that.

Reply:
@vyasr @wence- Yup, Both of you are correct.